### ISSUE TITLE
Feature request for AuRa: using `getPendingValidators()` getter instead of `InitiateChange` event

### ISSUE THREAD

**Issue by varasev** [poanetwork/parity-ethereum#49]
We agreed that we'll use the ~~`getValidators()`~~ [`getPendingValidators()`](https://github.com/poanetwork/pos-contracts/blob/1240a2e6ab098f18c868324fb2c22454417807d5/contracts/abstracts/ValidatorSetBase.sol#L192) getter instead of catching the `InitiateChange` event. Parity should just read the current validator set with ~~`getValidators()`~~ `getPendingValidators()` every block, and if it discovers that the set is changed in the contract, it launches changing of the set in the engine.

The [`finalizeChange()`](https://github.com/poanetwork/pos-contracts/blob/1240a2e6ab098f18c868324fb2c22454417807d5/contracts/abstracts/ValidatorSetBase.sol#L85) function (described in https://wiki.parity.io/Validator-Set.html) should be called by Parity after [`newValidatorSet()`](https://github.com/poanetwork/pos-contracts/blob/f7d080fbf5e16bd9b8cdb474bf9541bfe292c796/contracts/ValidatorSetAuRa.sol#L55) is called and the internal list of validators is updated in engine (this is default behavior for AuRa in Parity, so I think the code of the engine shouldn't be changed for that).

**UPDATE:** stroke out things below are already implemented by the contracts.

~~Parity must call [`newValidatorSet()`](https://github.com/poanetwork/pos-contracts/blob/1240a2e6ab098f18c868324fb2c22454417807d5/contracts/ValidatorSetAuRa.sol#L41) function by `SYSTEM_ADDRESS` once a week.~~

~~Since in AuRa the block time is determined (5 seconds), the `ValidatorSet` contract will keep track of elapsed time by itself, so there is no need to measure 1 week in Parity.~~

~~1 week = 120960 blocks~~

~~Parity must call the [`newValidatorSetCallable()`](https://github.com/poanetwork/pos-contracts/blob/1240a2e6ab098f18c868324fb2c22454417807d5/contracts/ValidatorSetAuRa.sol#L96) getter every block and check whether to call `newValidatorSet`.~~

~~So, if `newValidatorSetCallable()` returns `true`, the `newValidatorSet()` must be called at the end of the current block.~~

~~**Important note:** transaction of calling `newValidatorSet()` must be included in a block before transaction of calling `BlockReward.reward()`.~~

**Comment by varasev** [comment]
The description :arrow_up:  was updated:

1) I propose not to measure 1 week (duration of staking epoch) in Parity. We can do this in the contract.

2) It seems that Parity already uses `getValidators()` getter to verify that the internal list of validators is the same as the list in the contract. These lists should be identical. So, we need to use a getter with another name: `getPendingValidators()`.

I'll update the contract's code accordingly.

**Comment by DemiMarie** [comment]
@varasev can it be called *by* `BlockReward.reward()`?

**Comment by varasev** [comment]
@DemiMarie hmm, it's a good proposal. I'll check if this breaks something in the code. Thanks for the thought.

**Comment by DemiMarie** [comment]
@varasev The first thing that [`newValidatorSet()`][1] does is

```solidity
require(newValidatorSetCallable());
```
So we can just change it to:
```solidity
if (!newValidatorSetCallable()) {
    // do nothing
    return;
}
```
and just call it from [`BlockRewardAuRa.reward()`][2].

[1]: <https://github.com/poanetwork/pos-contracts/blob/1240a2e6ab098f18c868324fb2c22454417807d5/contracts/ValidatorSetAuRa.sol#L41>
[2]: <https://github.com/poanetwork/pos-contracts/blob/1240a2e6ab098f18c868324fb2c22454417807d5/contracts/BlockRewardAuRa.sol#L8>

**Comment by varasev** [comment]
Yes, I've modified the contracts in accordance with this: https://github.com/poanetwork/pos-contracts/blob/8d10a85ad87633a54a7444883b32687faf5b80bf/contracts/BlockRewardAuRa.sol#L38-L39

So, now we don't have to modify the AuRa source code for calling `newValidatorSet()`.

However, we still need to read `getPendingValidators()` instead of `InitiateChange` event.

I've updated the description above and changed the title.

**Comment by DemiMarie** [comment]
Fixed by #53

**Comment by afck** [comment]
Reopening; see https://github.com/poanetwork/parity-ethereum/issues/71#issuecomment-457167061.

**Comment by afck** [comment]
@DemiMarie, @varasev: I see this has been [reverted on `aura-pos`](https://github.com/poanetwork/parity-ethereum/commit/288362b79de406301bcbfb04f604cd6184090041). I guess it can be closed again, in favor of #71?

**Comment by varasev** [comment]
@afck Yes, I'm closing this.
